src/compiler/crystal/codegen/primitives.cr

```diff
@@ -291,10 +291,14 @@ class Crystal::CodeGenVisitor
 
   private def codegen_out_of_range(target_type : FloatType, arg_type : FloatType, arg)
     min_value, max_value = target_type.range
-    # arg < min_value || arg > max_value
-    or(
-      builder.fcmp(LLVM::RealPredicate::OLT, arg, float(min_value, arg_type)),
-      builder.fcmp(LLVM::RealPredicate::OGT, arg, float(max_value, arg_type))
+    # checks for arg being outside of range and not infinity
+    # (arg < min_value || arg > max_value) && arg != 2 * arg
+    and(
+      or(
+        builder.fcmp(LLVM::RealPredicate::OLT, arg, float(min_value, arg_type)),
+        builder.fcmp(LLVM::RealPredicate::OGT, arg, float(max_value, arg_type))
+      ),
+      builder.fcmp(LLVM::RealPredicate::ONE, arg, builder.fmul(float(2, arg_type), arg))
```

Comment:
Is there a reason not to compare infinity with equality?

Reply:
Equality with what? I just adapted the implementation from `Float#infinite?`.

Reply:
I mean comparing `arg` directly with an infinity constant. I imagine that for an optimizer is easier to deduce and simplify `arg != infinity && arg != - infinity`, rather than `arg != 2 * arg`

Reply:
Ah, I see. We don't happen to have such constants at hand, right?

Reply:
I think in IEEE infinity can be represented with multiple bit patterns. That's why you can't compare against infinity and there are specialized methods for this. So I think the suggestion is impossible to do.

Reply:
I think we can use Float32|64::Infinity with the `float` helper method. That is if fcmp compares infinity in an expected way.

Reply:
NaN has multiple values. But there are exactly two representation of infinity: Exponent all ones, fraction all zeros, sign for positive or negative.

Reply:
We need to compare against positive *and* negative infinity, but then it works.
```cr
     and(
        builder.fcmp(LLVM::RealPredicate::ONE, arg, float(arg_type.kind == :f32 ? Float32::INFINITY : Float64::INFINITY, arg_type)),
        builder.fcmp(LLVM::RealPredicate::ONE, arg, float(arg_type.kind == :f32 ? -Float32::INFINITY : -Float64::INFINITY, arg_type))
      )
```

I couldn't notice any performance difference in both implementations, though.

Reply:
I suppose we should keep the current one then, because it generates less code.

Reply:
Ok we can start with the current one. But more code might be easier to remove.